pilot/pkg/model/push_context.go

```diff
@@ -2460,3 +2460,8 @@ func (ps *PushContext) WaypointsFor(network, address string) []netip.Addr {
 func (ps *PushContext) WorkloadsForWaypoint(key WaypointKey) []WorkloadInfo {
 	return ps.ambientIndex.WorkloadsForWaypoint(key)
 }
+
+// ServicesForWaypoint returns all services associated with a given waypoint identified by it's network address
```

Comment:
What does 'network address' mean ? Can you also add some info on how this is supposed to be used - we may have waypoint implementations that are cross-namespace (or even cross cluster) - the result can be quite large. Having a 'waypoint for service' index is a clear use case - but not sure of the opposite.

Reply:
network address is a pretty commonly used key for the ambient index which is network name and IP address. Should be ok for cross-namespace and cross-cluster I would think. (fwiw, waypointKey attempts to abstract slightly to help with issues around different multi-ip scenarios such as dual-stack).

```go
type workloadsCollection struct {
	krt.Collection[model.WorkloadInfo]
	ByAddress        *krt.Index[model.WorkloadInfo, networkAddress]
	ByServiceKey     *krt.Index[model.WorkloadInfo, string]
	ByOwningWaypoint *krt.Index[model.WorkloadInfo, networkAddress]
}

type servicesCollection struct {
	krt.Collection[model.ServiceInfo]
	ByAddress        *krt.Index[model.ServiceInfo, networkAddress]
	ByOwningWaypoint *krt.Index[model.ServiceInfo, networkAddress]
}

// internal object used for indexing in ambientindex maps
type networkAddress struct {
	network string
	ip      string
}
```

ServicesForWaypoint is used to calculate/find the resources which a waypoint needs(see findWaypointResources). I think the alternative would be to loop through services, find there waypoint, then check if that waypoint is the one we're interested in which seems inefficient enough to justify the index to me. I will add some notes about this usage.